### ISSUE TITLE
Is it possible to render a partial without the parent scope?

### ISSUE THREAD

**Issue by aputinski** [wycats/handlebars.js#1032]
I'm trying to treat partials as "components" that only have access to variables that were passed in as attributes — is this possible?

**my-partial.hbs**
```html
<div class="my-partial">
  <!-- should only be displayed if it was passed in as an attribute -->
  {{#if name}}
    Hello, {{name}}
  {{/if}}
</div>
```

**my-app.js**
```javascript
const data = { name: 'Adam' }
```

**my-app.hbs**
```html
Hello, my name is {{name}} and is available in the template scope
{{> myPartial}}
{{> myPartial name="John"}}
```

### Actual output

```html
Hello, my name is Adam and is available in the template scope
<div class="my-partial>
  Hello, Adam
</div>
<div class="my-partial>
  Hello, John
</div>
```

### Desired output

```html
Hello, my name is Adam and is available in the template scope
<div class="my-partial"></div>
<div class="my-partial">
  Hello, John
</div>
```

**Comment by aputinski** [comment]
Here's my current workaround:

```javascript
Handlebars.registerHelper('component', (partial, options) => {
  const template = Handlebars.compile(Handlebars.partials[partial])
  const html = template(options.hash)
  return new Handlebars.SafeString(html)
})
```

```html
{{component "myPartial" name="Adam"}}
```

**Comment by kpdecker** [comment]
Passing a first argument to the partial call will render in its own context, independent of the parent. I.e. {{> myPartial newContext name="foo"}}

**Comment by rayHInstart** [comment]
Why is that not the default behavior? That was a rude shock when I found out.

**Comment by kpdecker** [comment]
How exactly would we default to providing a context from an arbitrary variable?

**Comment by rayHInstart** [comment]
I'm not sure I understand. If a set of key-value pairs were provided
that would form the object used as the context. For example if I had a
partial like this:

 {{> radioButton name="enabled" value="true" id="enableUser"
labelText="Enable User"}}

The object used as a context would be:
{
 name: "enabled",
 value: "true",
 id: "enableUser"
 labelText: "Enable User"
}

I don't know how it works now internally (I'm just a user).

**Comment by kpdecker** [comment]
@rayHInstart Those keys are added to the existing context to form a new context that has both sets of data, with the explicitly passed hash parameters given priority. This has a performance impact vs. some of the alternatives, but outside of that I'm not sure what was quite so rude or shocking. Glad to discuss any changed to behavior you'd like to see but please outline it fully as well as outline the issues that you've seen from the existing behavior.

**Comment by rayHInstart** [comment]
In my use case, I had a partial which used a common variable name,
something like

<div>
{{#if disabled}}Hi, I'm disabled{{/if}}
{{#unless disabled}}I'm super active!{{/unless}}
</div>

And I used it like this:
{{> superPartial }}

, expecting that since I did not explicitly pass a disabled into the
context, it would assume that disabled wasn't present. Unfortunately,
I had a completely unrelated variable named "disabled" in the template
object, so the partial kept rendering as if disabled were true, even
though I hadn't passed it in to the partial. I didn't expect this
behavior at all, and it took me quite a while to figure out what was
going on (couldn't easily find in the documentation a description of
this behavior). The partial is in a separate file, and I'm passing in
variables to the partial, so I assumed that the partial would always
have a completely different context than the parent. Since partials
are intended to be reused as parts of different templates, I would
think that their context would be compartmentalized away from the
parent context -- the partial shouldn't rely on the parent template
since that breaks encapsulation. I imagined the partial to be a
function which I was passing arguments to. I'm very glad that
Handlebars allows this capability if I tell it explicitly that I want
a new context for the partial, but to me it's counterintuitive that I
would have to tell it to do that. I think the expected behavior should
be that a partial always has its own context; if a user wants to
override and inherit from the parent context, there can be a flag to
do that, but it should be explicit.

**Comment by kpdecker** [comment]
Regarding documentation, the behavior is covered [here](http://handlebarsjs.com/partials.html) but as with all documentation, what makes sense to someone who already knows the code is not always the same for others.

> Will render the partial named `myPartial`. When the partial executes, it will be run under the current execution context.

I did not write the original implementation of this so I can't really comment on why the decisions were made as they were, but I can say that changing this is difficult. Many of these arguments are completely valid but like most issues filed with Handlebars of late, counter arguments can be made to equal degree, and for those cases "don't break existing code" generally wins.

I don't think it will be much overhead to add a flag that requires explicit contexts to be passed for partials which I can look at for this release, but we aren't going to be able to change the existing behavior. This does little to account for confusion of users just starting with the language, but would allow for your use case with out nasty temporary objects or similar hacks.

**Comment by rayHInstart** [comment]
Thanks Kevin. "Don't break existing code" is indeed a winning
argument. That is a flag I would definitely use :).

**Comment by kpdecker** [comment]
Released in 4.0.0

**Comment by arthwood** [comment]
To reference `local` variable specifically inside partial and not allow to "leak" from the outer context use:

```
@this.myVariable
```

**Comment by andreyvolokitin** [comment]
`this` doesn't work as a local context reference (`@this` doesn't work either). 

Is it possible to get a reference to the local context? If we have clashing name from some parent context, we could just reference a local context specifically inside a partial: `localCtx.clashingVar`. If it's not possible, how difficult is to implement this? Empty dumb context as the first partial argument just to "erase" parent context seems like littering in the code... An introduced flag is also seems not ideal, because sometimes you want to use parent by default (actually I relied on this until I encountered this issue). So, essentially it seems like it is best to allow choosing a context to use — parent by default, and local with some keyword reference. /cc @kpdecker
